File: Searching Algorithms/Binary_Search.cpp
//Time Complexity: O(logn)
#include<bits/stdc++.h>
using namespace std;
int main()
{

    int a[100],n,i,beg,end,mid,item;
    cout<<"Enter No. of Elements= ";
    cin>>n;
    cout<<"Enter Elements in Sorted Order=\n";
    for(i=1;i<=n;i++)
    cin>>a[i];

    cout<<"Enter Item you want to Search= ";
    cin>>item;

    beg=1;
    end=n;

    mid=(beg+end)/2;

    while(beg<=end && a[mid]!=item)
    {
        if(a[mid]<item)
            beg=mid+1;
        else
            end=mid-1;
        mid=(beg+end)/2;
    }
    if(a[mid]==item)
        cout<<"Data is Found at Location : "<<mid;
    else
        cout<<"Data is Not Found";

return 0;
}

File: Searching Algorithms/Python/binarysearch.py
def binary_search(item_list,item):
	first = 0
	last = len(item_list)-1
	found = False
	while( first<=last and not found):
		mid = (first + last)//2
		if item_list[mid] == item :
			found = True
		else:
			if item < item_list[mid]:
				last = mid - 1
			else:
				first = mid + 1	
	return found
	
print(binary_search([1,2,3,5,8], 6))
print(binary_search([1,2,3,5,8], 5))


"""
TIME COMPLEXITY

Best-case performance	        O(1)
Average performance	        O(log n)
Worst-case space complexity	O(1)

SPACE COMPLEXITY

Best-case performance	        O(1)
Average performance	        O(log n)
Worst-case space complexity	O(1)

"""
